### PR TITLE
Globe atmosphere + Style specification

### PR DESCRIPTION
**Pull request by karimnaaji**
## Atmosphere style specification

This PR extends the fog system to allow design of an atmosphere effect supported on both globe and mercator projections. It extends the `fog` style specification to account for three new values: `high-color`, `space-color` and `star-intensity`. The `star-intensity` allows for a parallax effect while rotating and panning the globe to increase the effect of 3d.

Webp net-gifmaker

## Changelist

- Add an `atmosphere` debug page with different preset to test the feature
- Add support for fog on g lobe with a fixed range, the fixed range value is chosen arbitrarily to give plausible rendering of a thin atmospheric layer over the globe.
- Consolidate frustum corner calculation and share it between atmosphere rendering and fog uniform upload
- Split atmosphere buffer from `GlobeSharedBuffer` so that it's not tied to globe rendering
- Remove clear color potentially using fog color, as this shader now covers the entire screen, it is unnecessary to dispatch an expensive clear
- Make sure that globe fixed fog range does not affect rendering of symbols and fog tile culling
- Make sure that heatmaps are not affected by globe fog fixed range, so that this layer can be rendered without being affected by fog opacity, as we get flickering artifacts otherwise
- Update render test baselines to account for the new changes
- Add style specification defaults using complex expressions so that `map.setFog({})` results in a nice default for the user
- Add a fake cubemap in the shader in order to create the parallax effect for stars

https://user-images.githubusercontent.com/7061573/162288266-c7804e9f-ed57-4db7-987d-65a0620c4f42.mov

## Style specification entries (including old ones):

- `color` alpha channel

color-alpha

  - `color` rgb channels

color

- `horizon-blend`

horizon-blend

- `high-color` alpha channel

sky-color-alpha

- `high-color` rgb channels

sky-color

- `space-color` alpha channel

space-color-alpha

- `space-color` rgb channels

space-color

- `star-intensity`

star-intensity

## Follow up to address after this PR

- Reduce frequency of globe specific uniforms when updating fog uniform, these values can be 1) calculated once per frame and 2) only be done when used along with globe
- ~~Fix horizon angle line visual artifact, refer~~ https://github.com/mapbox/gl-js-team/issues/410
- Add render tests specific to globe atmosphere
- Fix fill-extrusion fog contribution during transition

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [x] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'

### CODE REVIEW

File: src/style/properties.js [hunk 1]
```diff
@@ -759,7 +760,7 @@ export class Properties<Props: Object> {
             this.defaultTransitioningPropertyValues[property] =
                 defaultTransitionablePropertyValue.untransitioned();
             this.defaultPossiblyEvaluatedValues[property] =
-                defaultPropertyValue.possiblyEvaluate(({}: any));
+                defaultPropertyValue.possiblyEvaluate(defaultParameters);
```
Comment (mpulkki-mapbox) [inline src/style/properties.js:763]: Is this related to the atmosphere code?

File: src/shaders/atmosphere.fragment.glsl [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    vec3 position = vec3(p.xy * uv_scale + offset * u_viewport, p.z);
+
+    vec3 q = fract(position) - 0.5;
+    vec3 id = floor(position);
+
+    float random_visibility = step(random(id), u_star_density);
+    float circle = smoothstep(0.5 + u_star_intensity, 0.5, length(q));
+
+    return circle * random_visibility;
+}
```
Comment (mpulkki-mapbox) [inline src/shaders/atmosphere.fragment.glsl:34]: This produces a really nice looking starfield! Artefacts from distributing points on the cube (instead of uniform distribution on a sphere) are pretty much indistinguishable!

**Review by mpulkki-mapbox — COMMENTED** [review]
Looks really good! Are we supposed to have a default preset available when switching to globe view? Rest of the debug pages (e.g. terrain-debug.html) have no atmosphere set.

File: src/util/primitives.js [hunk 3]
```diff
@@ -85,6 +85,46 @@ class Ray {
...
+        const TR = vec3.transformMat4(TRClip, TRClip, invProj);
+        const BR = vec3.transformMat4(BRClip, BRClip, invProj);
+        const BL = vec3.transformMat4(BLClip, BLClip, invProj);
+
+        const horizonFromTopClip = 1.0 - (horizonFromTop / viewportHeight) * 2.0;
+
+        const horizonClipL = [-1, horizonFromTopClip, 1];
+        const horizonClipR = [1, horizonFromTopClip, 1];
+
+        const horizonL = vec3.transformMat4(horizonClipL, horizonClipL, invProj);
```
[the remark this answers is not in the document]
Comment (karimnaaji) [reply src/util/primitives.js:121]: Yes, good idea. I've updated to use `vec3.lerp` in latest. I also investigate whether we actually need two values here, because I believe we could simplify for a single uniform.

File: src/util/primitives.js [hunk 4]
```diff
@@ -85,6 +85,46 @@ class Ray {
     }
 }
 
+class FrustumCorners {
```
[the remark this answers is not in the document]
Comment (karimnaaji) [reply src/util/primitives.js:88]: There's at least one place in the code that doesn't need the frustum corners, where consolidating in `Frustum` would make this extraneous: https://github.com/mapbox/mapbox-gl-js/blob/main/src/ui/camera.js#L748 so I think it's ok to be split even though it's related.

File: src/style/properties.js (hunk 1 above)
Comment (karimnaaji) [reply src/style/properties.js:763]: This was necessary to fix an issue that manifested itself now that we introduced default parameters in a root style specification entry. Having a dummy parameter at creation allows further dynamic evaluation of the property values.

File: src/style/fog.js [hunk 5]
```diff
@@ -63,6 +79,12 @@ class Fog extends Evented {
             return;
         }
 
+        for (const name of Object.keys(styleSpec.fog)) {
+            if (fog && fog[name] === undefined) {
+                fog[name] = styleSpec.fog[name].default;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (karimnaaji) [reply src/style/fog.js:86]: Good catch, that's a bug, will fix that.

File: src/render/painter.js [hunk 6]
```diff
@@ -979,7 +986,27 @@ class Painter {
...
-                program.setFogUniformValues(context, fogUniformValues(this, fog, tileID, fogOpacity));
+                const tr = this.transform;
+                const viewMatrix = tr._camera.getWorldToCamera(tr.worldSize, 1.0);
+                const center = [tr.globeMatrix[12], tr.globeMatrix[13], tr.globeMatrix[14]];
+                const globeCenterInViewSpace = vec3.transformMat4(center, center, viewMatrix);
+                const globeRadius = tr.worldSize / 2.0 / Math.PI - 1.0;
+                const viewport = [
+                    tr.width * browser.devicePixelRatio,
+                    tr.height * browser.devicePixelRatio
+                ];
```
[the remark this answers is not in the document]
Comment (karimnaaji) [reply src/render/painter.js:997]: Yes, there are at least two improvements to make there that I mentioned in the description: 
>Reduce frequency of globe specific uniforms when updating fog uniform, these values can be 1) calculated once per frame and 2) only be done when used along with globe

because these are evaluated for every draw call, and could be evaluated once per frame instead. I'll fix that in this PR.

File: src/geo/projection/globe.js [hunk 7]
```diff
@@ -86,6 +86,11 @@ export default class Globe extends Mercator {
         return mat4.multiply(new Float64Array(16), tr.globeMatrix, decode);
     }
 
+    createFogTileMatrix(tr: Transform): Float64Array {
+        const matrix = mat4.identity(new Float64Array(16));
+        return mat4.multiply(matrix, matrix, tr.globeMatrix);
```
[the remark this answers is not in the document]
Comment (karimnaaji) [reply src/geo/projection/globe.js:91]: Simplified for `clone` in latest.

File: src/style/fog.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (karimnaaji) [reply src/style/fog.js]: Fixed in latest.

File: debug/atmosphere.html [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+        guiParams.enable = true;
+        guiParams.style = 'satellite-streets-v11';
+        guiParams.color = [220, 159, 159];
+        guiParams.highColor = [36, 92, 223];
+        guiParams.spaceColor = [0, 0, 0];
+        guiParams.spaceAlpha = 1.0;
+        guiParams.colorAlpha = 1.0;
+        guiParams.skyAlpha = 0.7;
+        guiParams.horizonBlend = 0.5;
+        guiParams.starIntensity = 0.15;
```
[the remark this answers is not in the document]
Comment (karimnaaji) [reply debug/atmosphere.html:71]: That should be fixed with latest.

**Comment by karimnaaji** [comment]
>Are we supposed to have a default preset available when switching to globe view?

No, there's no dependency between globe-view and atmosphere at the spec level. e.g. when setting globe-view we don't automatically enable the atmosphere. To improve UX, this is something to consider fixing at a higher level in Studio/when setting default styling, so that we're not adding dependencies between different style specification entries.

>Rest of the debug pages (e.g. terrain-debug.html) have no atmosphere set.

I've updated most of the pages in latest to use the default atmosphere.

There are two follow-ups to this PR:
- Reduce frequency of operations for https://github.com/mapbox/mapbox-gl-js/pull/11590#discussion_r847794446
- There's an issue that I've noticed on fill-extrusion on globe during the transition. I'll  address that separately:

https://user-images.githubusercontent.com/7061573/163053674-7f9a8344-7f83-473e-a324-5624c79cc961.mov

Otherwise please take another look as this is ready for another review, I think I've addressed all of the other comments.

File: debug/atmosphere.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (mpulkki-mapbox) [reply debug/atmosphere.html:71]: Still seeing some issues. This is just a minor thing as all UI controls work!

https://user-images.githubusercontent.com/55925868/163238266-0d61ec61-a62d-4ad5-ae82-c37c13e6ac5d.mp4

File: src/geo/projection/globe.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (mpulkki-mapbox) [reply src/geo/projection/globe.js]: `mat4.clone()` returns a matrix using a Float32Array. This might have undesirable side effects?

File: src/style/properties.js (hunk 1 above)
Comment (mpulkki-mapbox) [reply src/style/properties.js:763]: okay thanks for the clarification! :+1:

File: src/shaders/atmosphere.fragment.glsl [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+void main() {
+    highp vec3 dir = normalize(v_ray_dir);
+
+#ifdef PROJECTION_GLOBE_VIEW
+    float globe_pos_dot_dir = dot(u_globe_pos, dir);
+    highp vec3 closest_point_forward = abs(globe_pos_dot_dir) * dir;
+    float norm_dist_from_center = length(closest_point_forward - u_globe_pos) / u_globe_radius;
+
+    if (norm_dist_from_center < 1.0) {
+        gl_FragColor = u_color;
```
Comment (mpulkki-mapbox) [inline src/shaders/atmosphere.fragment.glsl:45]: early exit here?

File: src/shaders/atmosphere.fragment.glsl [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    float horizon_angle = horizon_angle_mercator;
+#endif
+
+    // Normalize in [0, 1]
+    horizon_angle /= PI;
+
+    // exponential curve
+    // [0.0, 1.0] == inside the globe, > 1.0 == outside of the globe
+    // https://www.desmos.com/calculator/l5v8lw9zby
+    float t = exp(-horizon_angle * pow(u_fadeout_range, -1.0));
```
Comment (mpulkki-mapbox) [inline src/shaders/atmosphere.fragment.glsl:79]: is this an optimization to write `pow(x, -1)` instead of `1/x`?

Comment (karimnaaji) [reply src/shaders/atmosphere.fragment.glsl:79]: Not an optimization, that's an overlook from converting the function from desmos, thanks for pointing that out I'll update it!

Comment (karimnaaji) [reply src/shaders/atmosphere.fragment.glsl:79]: Also, I switched that function to prefer `exp` over an `sqrt` based function, so that the effect is smoother, you can notice that the `sqrt` based function that we used in `main` (in red) results in an abrupt change when crossing `y=0` due to clamping:

Screenshot from 2022-04-13 11-29-45

Comment (mpulkki-mapbox) [reply src/shaders/atmosphere.fragment.glsl]: yes I noticed! That is a better option :+1:

File: debug/atmosphere.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (karimnaaji) [reply debug/atmosphere.html:71]: Oh ok I see what you mean now, you'd like the UI to update based on the preset values! I'm not sure if that's possible with this GUI library but I'll give it a look.

File: src/geo/projection/globe.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (karimnaaji) [reply src/geo/projection/globe.js]: Yes, good catch. Will revert from clone and simplify.

**Review by mpulkki-mapbox — APPROVED** [review]
LGTM!

File: debug/atmosphere.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (karimnaaji) [reply debug/atmosphere.html:71]: Using `.listen()` did the trick to refresh the values: https://github.com/mapbox/mapbox-gl-js/pull/11590/commits/73566dc190d71722968eb1d015de04d3b64a47c3
